Mobile - Slash inserter

geriux opened the pull request:
Fixes: https://github.com/wordpress-mobile/gutenberg-mobile/issues/2620

<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

## Description

This PR adds the `/` (slash) key shortcut to add a block on mobile.

`Gutenberg Mobile PR` -> https://github.com/wordpress-mobile/gutenberg-mobile/pull/3250
`WordPress iOS PR` -> https://github.com/wordpress-mobile/WordPress-iOS/pull/16064
`WordsPress Android PR` -> https://github.com/wordpress-mobile/WordPress-Android/pull/14245

**Pending issues**:
- [x] Update UI styles:
  - Add a background color to the highlighted item: `$gray-light` (light mode) and `$neutral-dim-dark` (dark mode).
  - <img src="https://user-images.githubusercontent.com/4885740/111274586-75a3f480-8635-11eb-9434-fc11c7e3f49f.png" width=300 />
- [x] ~Hide feature behind a feature flag for DEV~ once this PR is approved it'll ship to production so no DEV flag needed.
- [x] Issues with [Android and the onEnter event](https://github.com/WordPress/gutenberg/pull/29683/#issuecomment-799146552). Here's a [PR](https://github.com/WordPress/gutenberg/pull/30368) with a fix for this but it won't be a blocker to ship this feature.
- [x] React warning `React.createElement: type is invalid -- expected a string (for built-in components)`
- [x] Add E2E tests/sanity tests
- [x] Update CHANGELOG notes

## How has this been tested?
- Open the app
- Within a `Paragraph` block
- Type the `/` character following by any block name
- Expect to see the block list suggestion above the keyboard
- Once the block you typed is highlighted tap on the enter key
- Expect to see the block added

## Screenshots 

iOS Light | iOS Dark
-|-
<kbd><img src="https://user-images.githubusercontent.com/4885740/117946935-08000600-b310-11eb-8e97-927e48ab7a57.gif" width="250" /></kbd> | <kbd><img src="https://user-images.githubusercontent.com/4885740/117946928-05051580-b310-11eb-9d03-4143d59acdfb.gif" width="250" /></kbd> 

Android Light | Android Dark
-|-
<kbd><img src="https://user-images.githubusercontent.com/4885740/117947029-1b12d600-b310-11eb-86bb-2fa49a5b4f13.gif" width="250" /></kbd> | <kbd><img src="https://user-images.githubusercontent.com/4885740/117947014-1817e580-b310-11eb-806b-a4b1be2a9be2.gif" width="250" /></kbd> 

## Types of changes
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

mkevins committed:
Crudely fork web implementation to remove invariant violations

Commit:
WIP - Better UI for testing. Reset Autocomplete on delete content

mkevins committed:
Use a synthetic event to bind onEnter to the autocompletion

Commit:
Mobile - Move onEnter logic to native Rich Text

mkevins committed:
Simplify event conformance in native RichText

mkevins committed:
Remove native variant of block-editor autocomplete

This adds the block completer

mkevins committed:
Use Platform variant to wrap text on native, leaving web unchanged

mkevins committed:
Extract autocomplete functions

mkevins committed:
Merge autocomplete variants, and create an autocomplete ui variant

mkevins committed:
WIP - add Popover library and begin experimenting

mkevins committed:
Move optional from this to function

Commit:
WIP - Popover changes

mkevins committed:
Extract default useItems hook from autocomplete ui

mkevins committed:
Rename extracted ui for consistency

Commit:
Remove Popover library. Add UI for the slash inserter selector

Commit:
Remove lodash usage

Commit:
Fix preventDefault on Android

mkevins committed:
Simplify customEditableOnKeyDown stubs

mkevins committed:
Rename autocompletion slot fill and move to mobile components

mkevins committed:
Reset scroll position when autocomplete items change

enejb commented on packages/components/src/autocomplete/style.native.scss:
```diff
@@ -0,0 +1,61 @@
+.item {
+	flex-direction: row;
+	align-items: center;
+	padding-top: 5px;
```
Could this be rewritten as `padding: 5px 12px 8px 8px;` ?

geriux replied:
Of course, thanks for the suggestion!

enejb commented:
I tested this briefly via https://github.com/wordpress-mobile/WordPress-iOS/pull/16064
And it worked really well. Nice work!

mkevins committed:
Adjust styles for BlurView

(for iOS - needs Android changes)

mkevins committed:
Merge branch 'trunk' into rnmobile/feature/slash-inserter

mkevins committed:
Update autocompletion item container view styles for Android

Commit:
Update autocomplete styles

Commit:
Merge branch 'trunk' into rnmobile/feature/slash-inserter

Commit:
Fix order of imports

mkevins committed:
Put feature behind __DEV__ flag

mkevins committed:
Merge branch 'trunk' into rnmobile/feature/slash-inserter

Commit:
Remove mobile wrapper since it no longer uses the label attribute to render the block name

Commit:
Add missing line break

Commit:
Update active item styles

Commit:
Revert "Put feature behind __DEV__ flag"

This reverts commit 528a83fde6ca53b7d0126aea077f97ab047627fd.

Commit:
Merge branch 'trunk' into rnmobile/feature/slash-inserter

geriux commented:
Hey @iamthomasbishop 👋 I've made a new [iOS build](https://github.com/wordpress-mobile/WordPress-iOS/pull/16064#issuecomment-796711814) with the latest style changes, can you please check it out when you can? Thanks!

iamthomasbishop commented:
@geriux This is so great to see in action, nice work! I only have a couple of minor suggestions, and one thought that I will drop at the end for future consideration.

- The entry of the UI is a bit too subtle, making it easy to miss. Would it be possible to start the toolbar at the same position as the Block Toolbar and have it slide up over ~200-300ms? I can gather some specific transition properties/values if necessary, but perhaps we can reuse what we have for entry/exit of BottomSheets 🤔
- `$gray-light` for the highlighted item’s background is a little bit too dim in light mode. Can we try `$neutral-dim-light` for that background? If that as a neutral background clashes with the more saturated icon color, let’s try a step lighter than the current `$gray-light`, which I believe would be `$gray-lighten-30`. (Note: `$neutral-dim-dark` looks great in dark mode!)
- Do you think we should change the placeholder label on empty Paragraph blocks to match the web? The label on web is “Start writing or type / to choose a block”, but here are some additional options:
	- ”Start writing or type / to add content”
	- ”Start writing or type / to add a block”
	- ”Start writing or type /“

**Something to consider in the future**

I hadn’t realized until I was stress testing this that it only works on Paragraph blocks, whereas for some reason I was expecting it to work on any (empty) RichText-based block (if I wanted to go between Heading <> Paragraph, for example). This obviously matches the web UI’s behavior, so it’s not worth diverging at this point, but it would be nice to consider for the next iteration, and to align this with the preliminary work @illusaen is doing on [Block Transforms](https://github.com/WordPress/gutenberg/pull/28453) for mobile.

mkevins committed:
Merge branch 'trunk' into rnmobile/feature/slash-inserter

mkevins committed:
Add slash-inserter e2e test

mkevins committed:
Update light-mode active background color for slash complete option

mkevins commented:
> * `$gray-light` for the highlighted item’s background is a little bit too dim in light mode. Can we try `$neutral-dim-light` for that background? If that as a neutral background clashes with the more saturated icon color, let’s try a step lighter than the current `$gray-light`, which I believe would be `$gray-lighten-30`. (Note: `$neutral-dim-dark` looks great in dark mode!)

I've updated this to use `$gray-lighten-30`, since I think that you are right about the slight color "clash". Here they are for comparison:

Original (`$gray-light`) | neutral dim light (actually `$light-dim`) | `$gray-lighten-30`
-|-|-
![active-background-color-gray-light](https://user-images.githubusercontent.com/8507675/113376130-8ad09100-9360-11eb-9ed6-6a03fae05933.png) | ![active-background-color-light-dim](https://user-images.githubusercontent.com/8507675/113376135-8c9a5480-9360-11eb-926f-f7010d4a0213.png) | ![active-background-color-gray-lighten-30](https://user-images.githubusercontent.com/8507675/113376133-8c01be00-9360-11eb-8eb7-c2e3e0c84d62.png)

Note: I didn't see a $neutral-light-dim, but there was a neutral light "section" in the colors file, with `$light-dim` as a variable. Please let me know if you had a different one in mind.

> * Do you think we should change the placeholder label on empty Paragraph blocks to match the web? The label on web is “Start writing or type / to choose a block”, but here are some additional options:
>   
>   * ”Start writing or type / to add content”
>   * ”Start writing or type / to add a block”
>   * ”Start writing or type /“

I like this idea! One thing to note, this has very recently changed on web in [this PR](https://github.com/WordPress/gutenberg/pull/29611). Not sure if that weighs in on our choice here :thinking:

iamthomasbishop commented:
> Note: I didn't see a $neutral-light-dim, but there was a neutral light "section" in the colors file

@mkevins Ahh, my bad. `light-dim` is the one, but I should've known that'd be too heavy. Can we try `light-ultra-dim` in light mode? That should be quite a bit better.

> One thing to note, this has very recently changed on web in this PR. Not sure if that weighs in on our choice here 🤔

Nice catch! I hadn't seen that one 😄 We can probably align with that because it looks like it was merged. But let's make sure we get a couple of set of eyeballs on it to review the wording — I don't think it's _critical_ that we match that exactly on mobile.

Commit:
Update placeholder text to hint the slash inserter

Commit:
Slash inserter animation and active item color update

Commit:
Styles cleanup

geriux commented:
Hey @iamthomasbishop 👋 thank you for the feedback!

> The entry of the UI is a bit too subtle, making it easy to miss. Would it be possible to start the toolbar at the same position as the Block Toolbar and have it slide up over ~200-300ms? I can gather some specific transition properties/values if necessary, but perhaps we can reuse what we have for entry/exit of BottomSheets 🤔

I added an entry/exit animation in the latest iOS build, please let me know what you think =)

> Do you think we should change the placeholder label on empty Paragraph blocks to match the web? The label on web is “Start writing or type / to choose a block”, but here are some additional options:

This is also included in the latest build, in my opinion, it is great to have this hint to teach users how to use the slash inserter.

> Can we try light-ultra-dim in light mode? That should be quite a bit better.

I updated this in the build but here are some screenshots:

iOS | Android
-|-
<kbd><img src="https://user-images.githubusercontent.com/4885740/113711519-5274de00-96e5-11eb-8311-3ef4555e1fe1.png" width="250" /></kbd> | <kbd><img src="https://user-images.githubusercontent.com/4885740/113711532-556fce80-96e5-11eb-8ed2-cdfd2d225cad.png" width="250" /></kbd> 

Let us know what you think, thanks!

Commit:
Pass isVisible to useCallback

enejb commented on packages/components/src/autocomplete/style.native.scss:
```diff
@@ -0,0 +1,73 @@
+}
+
+.activeTextDark {
+	text-decoration: underline;
```
This line seems redundant. Since it is already part of the activeText.

mkevins replied:
Nice catch! Fixed with https://github.com/WordPress/gutenberg/pull/29772/commits/4d184ac9e6ab1f4068e1dfa417ce6f8a20c44f0b

enejb commented on packages/components/src/autocomplete/style.native.scss:
```diff
@@ -0,0 +1,73 @@
+.wrapper {
```
I think we should start following the new BEM stylesheet pattern in the new document.

mkevins replied:
This sounds interesting. Is there an example of BEM usage on the RN side of things? I know there is already some "magic" happening between kebab-case and camelCase things via our scss transpilation (since JS can't have variables / property names with `-`), but curious about the _new_ pattern you mention.

geriux replied:
Thanks @enejb for the suggestion, we started this feature before deciding about using the BEM pattern but we'll look into it.

geriux replied:
Updated it here https://github.com/WordPress/gutenberg/pull/29772/commits/47bd27e56ffc0a1d5c2ab5da36735603457e0201

enejb commented on packages/react-native-editor/__device-tests__/slash-inserter.test.js:
```diff
@@ -0,0 +1,28 @@
+import { isAndroid } from './helpers/utils';
+
+describe( 'Gutenberg Editor tests for slash inserter', () => {
+	it( 'should insert an image block with enter', async () => {
```
Can we be more specific here. 
```suggestion
	it( 'should insert an image block with "/img" + enter', async () => {
```

enejb commented:
Nice work here! I love how it is coming together. 

Re the BEM note. I think it might be worthwhile to the changes now since we are trying to move to the new format.

I am not sure but would it make sense to add a readme for the new component as well? So that it would be easier for others to implement similar features in the future.

iamthomasbishop commented:
> I added an entry/exit animation in the latest iOS build, please let me know what you think =)

@geriux That animation definitely helps. Can we shorten up the duration of the transitions in/out by any chance? It looks like it's 300ms/150ms (in/out), so maybe we can try 200/100.

> This is also included in the latest build, in my opinion, it is great to have this hint to teach users how to use the slash inserter.

After seeing this in action, I think the label may be a bit too long. Maybe for now, we can probably match the [recent update](https://github.com/WordPress/gutenberg/pull/29611) from Core that @mkevins mentioned, which is `Type / to choose a block`.

> I updated this in the build but here are some screenshots:

That definitely looks a little bit better, but I think the colors are still a little bit off and could be refined to be a little more resilient. I did some quick tinkering using a variety of imagery behind the surface and came up with some refined colors in light mode. From what I can gather from your screenshots, this would amount to 3 changes (for light mode only):

- Highlighted button background color would become `$gray-darken-20` with an `alpha` of `.1`
- The text label/icon color of the highlighted button would become `$gray-dark`
- Non-highlighted button text/icons would become `$gray-darken-30`

Which altogether looks like this:

<a href="https://user-images.githubusercontent.com/1200432/113784140-827aac00-96fa-11eb-9b81-6f997b5a4b7a.png"><img src="https://user-images.githubusercontent.com/1200432/113784140-827aac00-96fa-11eb-9b81-6f997b5a4b7a.png" width="2395"></a>

One other question for you, based on your screenshot of Android above. I noticed the background color of the toolbar is white (with a `border-top`) on Android — aren't we using the same base color (which I think is an off-white `#f9f9f9` and no border or very close to it on iOS) for Compact Notices on both platforms? 

This also got me thinking, what if we were to use white as the base color instead, so that in its default state it blends in with the canvas? We should probably stick with the standard background style we're using bc it aligns with the notices, but this does blend in pretty nicely, and fwiw I could see us re-using this more "camouflaged" material in other places (like the block toolbar, eventually) 🤔 Any thoughts or opinions @geriux @mkevins ? Obviously we can stick with the surface we have currently to get this out the door, but I wanted to throw it out there 😄 

<a href="https://user-images.githubusercontent.com/1200432/113785175-3e88a680-96fc-11eb-8667-d13041d735b4.png"><img src="https://user-images.githubusercontent.com/1200432/113785175-3e88a680-96fc-11eb-8667-d13041d735b4.png" width="1351"></a>

mkevins committed:
Adjust autocompleter animation timing

mkevins committed:
Remove redundant text decoration

mkevins committed:
Adjust colors

mkevins committed:
Use shorter writing prompt

This is taken from #29611

mkevins committed:
Make test description more specific

Co-authored-by: Enej Bajgoric <enej.bajgoric@automattic.com>

Commit:
Styles update

geriux commented:
Hey @iamthomasbishop 👋  thanks for the feedback!

> One other question for you, based on your screenshot of Android above. I noticed the background color of the toolbar is white (with a border-top) on Android — aren't we using the same base color (which I think is an off-white #f9f9f9 and no border or very close to it on iOS) for Compact Notices on both platforms?

It was a mistake on our side, we were indeed using a white background color and not the `Notice` background color. I updated the background color also removed the border on the top.

> This also got me thinking, what if we were to use white as the base color instead, so that in its default state it blends in with the canvas? We should probably stick with the standard background style we're using bc it aligns with the notices, but this does blend in pretty nicely, and fwiw I could see us re-using this more "camouflaged" material in other places (like the block toolbar, eventually)

I like that approach but the only way to achieve a full white background using the blur feature is to change the current mode we are using from `prominent` to `light` or `regular` and it only works for light mode:

## Using a white background for light mode

### iOS Light
https://user-images.githubusercontent.com/4885740/113895005-d358d600-97c8-11eb-8d6a-df726d6049d3.mov

### iOS Dark
https://user-images.githubusercontent.com/4885740/113895059-e10e5b80-97c8-11eb-9f9f-d91aa83f273b.mov

### Android Light
https://user-images.githubusercontent.com/4885740/113896484-4282fa00-97ca-11eb-8c9d-13c49c5bf979.mov


### Android Dark
https://user-images.githubusercontent.com/4885740/113896505-46af1780-97ca-11eb-85f5-1c5cbbe1956c.mov

## Using the current approach but with the latest style changes

Thanks to @mkevins for updating pretty much all of them,

iOS Light | iOS Dark
-|-
<kbd><img src="https://user-images.githubusercontent.com/4885740/113890951-ebc6f180-97c4-11eb-90c9-53d935dfe5b5.png" width="250" /></kbd> | <kbd><img src="https://user-images.githubusercontent.com/4885740/113890962-ee294b80-97c4-11eb-87c8-cbffa5f2aef9.png" width="250" /></kbd> 

Android Light | Android Dark
-|-
<kbd><img src="https://user-images.githubusercontent.com/4885740/113891160-1add6300-97c5-11eb-8b8c-6944bd515c14.png" width="250" /></kbd> | <kbd><img src="https://user-images.githubusercontent.com/4885740/113891171-1ca72680-97c5-11eb-873a-acf70e5f1468.png" width="250" /></kbd> 

Let us know what you think, we can make a new build once we decide if we keep the current "Solid" background or go with the `white` background.

Thanks!

iamthomasbishop commented:
> It was a mistake on our side, we were indeed using a white background color and not the Notice background color. I updated the background color also removed the border on the top.

Ah okay, thank you!

> I like that approach but the only way to achieve a full white background using the blur feature is to change the current mode we are using from prominent to light or regular and it only works for light mode:

Thanks for humoring me 😄  Let’s just stick with the current w/ updated styling.

I wouldn’t mind seeing one more test build — the updates text/icon colors in light mode feel a tad dark, but I err’ed on the dark side to ensure there is still enough contrast in scenarios where there is a colorful image or other busy content behind it. We’ll definitely want to darken up all of the toolbar buttons to be consistently colored at some point, but that’s for a different day 😄

Commit:
Merge branch 'trunk' into rnmobile/feature/slash-inserter

Commit:
Merge branch 'trunk' into rnmobile/feature/slash-inserter

geriux commented:
Hey @iamthomasbishop 👋 

> I wouldn’t mind seeing one more test build

The iOS build is now available =)

 ~I'm trying to make an Android one as well.~

**Edit**: [Android build](https://101966-9306568-gh.circle-artifacts.com/0/Artifacts/WordPress-pr-14245-build-101966.apk) is ready too.

Let us know what you think. Thanks!

Commit:
Mobile: Update default text placeholder, fix autocompleter position in Android

Commit:
Merge branch 'trunk' into rnmobile/feature/slash-inserter

# Conflicts:
#	packages/components/src/autocomplete/index.js

geriux commented:
Hey @iamthomasbishop 👋 

Thanks for testing the builds again and [providing feedback](https://github.com/wordpress-mobile/WordPress-iOS/pull/16064#issuecomment-821541499).

> Issue: Blocking toolbar on Android

I created a new build for both platforms and it fixes the issue you mentioned for Android, could you check it out? Thanks!

> Proposals
>I have often felt the "italics" icon that we are using (from the WP icon set aka "G2") was a little confusing, but it hasn't caused any real conflicts until now, because the icon tends to look like a "slash", which is obviously a potential issue considering the "/" character of the placeholder/writing prompt is close enough in proximity to the "italic" icon in the block toolbar.

Yeah, I agree that since it's too close and the icon looks like the `/` character it will cause confusion 😅 

> I found at least one issue upon a quick GitHub search, and think it'd be worth proposing some updates to the icon in core, but in the meantime I propose that we use a custom icon on mobile

I feel like the proposal to core would be the ideal approach, currently, we use the same icons, if we were to use a different one for mobile we'd need to duplicate the `italics` code to specify just a different icon. Its totally fine if we do it but just in case we can just wait a little and aim for the first option.

> Proposal: refined writing prompt

Here I went for option A, since adding the icon inside the placeholder is not something trivial and definitely needs investigation, you have this change in the latest builds as well. Can you please check it out and let me know what you think?

Thanks again for your time testing this feature and making it even better with your feedback!

iamthomasbishop commented:
> I created a new build for both platforms and it fixes the issue you mentioned for Android, could you check it out? Thanks!

I'm not seeing a new build for iOS but I tested on Android and you've fixed the blocker, thank you! 

One additional refinement, but we can spin out into a separate ticket if necessary because you've addressed the blocking issue: I'd like to bump the block position upward a bit more (~48px, on Android) so that the inline toolbar buttons (mover icons, settings and ellipsis icons) is visible and aligns to ~4-8px from the top edge. Here's a quick example:

<a href="https://user-images.githubusercontent.com/1200432/116442901-a83a3300-a818-11eb-8db5-029cb89d6158.png"><img src="https://user-images.githubusercontent.com/1200432/116442901-a83a3300-a818-11eb-8db5-029cb89d6158.png" width="1203"></a>

> I feel like the proposal to core would be the ideal approach, currently, we use the same icons, if we were to use a different one for mobile we'd need to duplicate the italics code to specify just a different icon. Its totally fine if we do it but just in case we can just wait a little and aim for the first option.

Agreed this is ideal in the long-term. Moving forward, I'll make sure we get a proposal up on the core side to refine the icon. In the meantime, I'm curious about the effort required to replace the italics icon with a custom one on our side (temporarily, assuming a proposal to update the icon on core is accepted). Is this something that when applied in once place would cascade to all other instances of that icon in the editor, or would we need to manually implement this for every instance?

If it's the former (a matter of applying the new icon once and it simply cascading across the editor), then I would prefer to move forward with that approach. On the other hand, if it will require us to implement manually in a bunch of locations across the editor, that seems a bit fragile and time-consuming, so we can just roll with what we have now.

> Here I went for option A, since adding the icon inside the placeholder is not something trivial and definitely needs investigation, you have this change in the latest builds as well. Can you please check it out and let me know what you think?

Looks good! I want to get another set of design eyes on it because it's a relatively consequential change in terms of block discovery. // cc @kyleaparker How do you feel about changing the placeholder text label on empty paragraph blocks (currently is "Start writing...") to "Tap (+) or type / to add blocks"? It looks like this:

<a href="https://user-images.githubusercontent.com/1200432/116447892-19c8b000-a81e-11eb-89e3-83298b8e2778.png"><img src="https://user-images.githubusercontent.com/1200432/116447892-19c8b000-a81e-11eb-89e3-83298b8e2778.png" width="907"></a>

geriux commented:
Hey @iamthomasbishop 👋 

> I'm not seeing a new build for iOS but I tested on Android and you've fixed the blocker, thank you!

Oh that's weird, on iOS the usual comment with the build links didn't work. Glad to know that we don't have a blocker on Android anymore 🎉 

> One additional refinement, but we can spin out into a separate ticket if necessary because you've addressed the blocking issue: I'd like to bump the block position upward a bit more (~48px, on Android) so that the inline toolbar buttons (mover icons, settings and ellipsis icons) is visible and aligns to ~4-8px from the top edge. Here's a quick example:

That refinement would be great, I can create a separate issue with your proposal.

> If it's the former (a matter of applying the new icon once and it simply cascading across the editor), then I would prefer to move forward with that approach. 

Sure! We can do this, we only need to change one file so if you can please give us an SVG with the new icon we can make the change.

Thanks!

iamthomasbishop commented:
> That refinement would be great, I can create a separate issue with your proposal.

Great, thank you!

> Sure! We can do this, we only need to change one file so if you can please give us an SVG with the new icon we can make the change.

Okay, awesome! I'll prepare the SVG today. Thanks, @geriux!

iamthomasbishop commented:
@geriux I added the SVG to our design docs under the [Mobile GB tab](https://zeroheight.com/226850145/p/345d49-icons/b/78279c) of the Icons page, named `icon-italics-alt`. Note that it's _slightly_ different from my previous example, where the icon was a bit more slanted. I decided it'd be best to align the angle more closely with the current icon, but did keep the horizontal top/bottom lines. Here are some examples in context:

<a href="https://user-images.githubusercontent.com/1200432/116613312-215f8600-a8fe-11eb-8969-5f0ae1cb048c.png"><img src="https://user-images.githubusercontent.com/1200432/116613312-215f8600-a8fe-11eb-8969-5f0ae1cb048c.png" width="517"></a>

<a href="https://user-images.githubusercontent.com/1200432/116613347-2f150b80-a8fe-11eb-8855-32949bb78645.png"><img src="https://user-images.githubusercontent.com/1200432/116613347-2f150b80-a8fe-11eb-8855-32949bb78645.png" width="1144"></a>

Note: Let me know if there is anything special we need to do in terms of formatting or refining the SVG and I'll take care of it.

mkevins committed:
Use I with serif for italic format library icon

This is to distinguish the character from the slash character to avoid
confusion with the slash inserter (especially with the new writing
prompt that says to type "/").

Commit:
Mobile - Add Slash inserter e2e tests

Commit:
Merge branch 'rnmobile/feature/slash-inserter' of github.com:WordPress/gutenberg into rnmobile/feature/slash-inserter

kyleaparker commented:
>Looks good! I want to get another set of design eyes on it because it's a relatively consequential change in terms of block discovery. // cc @kyleaparker How do you feel about changing the placeholder text label on empty paragraph blocks (currently is "Start writing...") to "Tap (+) or type / to add blocks"? It looks like this:

@iamthomasbishop do you think the tap (+) makes the new tooltip redundant? I'm also worried it might make people think they need to go to the inserter to start writing. There was something nice about the simplicity to the "Start writing..." prompt for new users, where this adds a little more complexity. I think it may be worth exploring other ways to communicate the / inserter, especially since it's a power feature I'm not sure it needs to be part of the placeholder

Commit:
Merge with trunk

Commit:
Unify e2e files

iamthomasbishop commented:
@kyleaparker I definitely agree that "start writing..." is lighter, more approachable and aesthetically pleasing, my concern would be losing the aided discoverability that this provides. If we're okay with losing that (at least temporarily) then maybe we can just keep the current "Start writing..." label, and defer any changes to when explore this as part of the wider block discovery work (likely starting in the next 3-4 weeks). @geriux @mkevins how does that sound?

(As an aside: Fwiw, my opinion on how much I'd consider "slash commands" a power-user thing has shifted over the past 2-3 years as it has become a more prominent pattern in various popular apps (Slack, Notion, etc), to the point where it's sort of in the gray area; especially considering it's a pretty easy hurdle for users to clear and interaction to repeat. But we can save that discussion for another day 😄 )

geriux commented:
> @geriux @mkevins how does that sound?

It sounds good! So we'll change it back to the current label. Do we still want to make the `italic` icon change? Or should we revert that as well?

iamthomasbishop commented:
> Do we still want to make the italic icon change? Or should we revert that as well?

@geriux  Let's continue with the proposed "italics" icon. We will give the text label on the block a closer inspection in the coming weeks.

geriux commented:
> @geriux Let's continue with the proposed "italics" icon. We will give the text label on the block a closer inspection in the coming weeks.

Awesome! Thanks! I'll update the label and make a build to test 🚀

Commit:
Revert placeholders change

geriux commented:
Builds are ready:

- [iOS](https://appcenter-filemanagement-distrib2ede6f06e.azureedge.net/d04890fc-640a-43c2-8231-5acab5653b48/WordPress%20Alpha.ipa?sv=2019-02-02&sr=c&sig=waCDy0AnGgPbg%2B9e%2FGnOOaGmzKJdNLBhlRysHrYKRzA%3D&se=2021-05-01T18%3A42%3A14Z&sp=r)
- [Android](https://104338-9306568-gh.circle-artifacts.com/0/Artifacts/WordPress-pr-14245-build-104338.apk)

kyleaparker commented:
> I definitely agree that "start writing..." is lighter, more approachable and aesthetically pleasing, my concern would be losing the aided discoverability that this provides. If we're okay with losing that (at least temporarily) then maybe we can just keep the current "Start writing..." label, and defer any changes to when explore this as part of the wider block discovery work (likely starting in the next 3-4 weeks). @geriux @mkevins how does that sound?

This sounds good to me! I definitely think we want to introduce people to the slash inserter, but I think it's worth doing some more design exploration on potential options before we make any changes here

iamthomasbishop commented:
> This sounds good to me! I definitely think we want to introduce people to the slash inserter, but I think it's worth doing some more design exploration on potential options before we make any changes here

Ok cool, it sounds like we're in agreement then 👍 Stoked to get this out the door and in users' hands!

geriux commented:
> > This sounds good to me! I definitely think we want to introduce people to the slash inserter, but I think it's worth doing some more design exploration on potential options before we make any changes here
> 
> Ok cool, it sounds like we're in agreement then 👍 Stoked to get this out the door and in users' hands!

Awesome! @iamthomasbishop  Does that mean we have the ✅?

geriux commented:
Hey @ellatrix 👋 We added support for the `/` autocomplete on mobile and we separated the base code into some files from the `autocomplete` file to share some parts of the code, I double-checked that works correctly on the web editor but it'd be great if you can give it a look. Thanks!

iamthomasbishop commented:
> Does that mean we have the ✅?

@geriux I just took another look at the test build on iOS and iPad OS to make sure, and it's looking good. Let's 🚢 it! 👏

ellatrix commented:
@geriux Sounds fine. Worth nothing that in the future we'd like to get rid of this component and create something more integrated with rich text. As long as we're not exposing new APIs here I'm fine with it.

ellatrix commented on packages/edit-post/src/components/layout/index.native.js:
```diff
@@ -15,6 +15,7 @@ import {
 	HTMLTextInput,
 	KeyboardAvoidingView,
 	NoticeList,
+	AutocompletionItemsSlot,
```
It looks like these are new APIs. Could they be marked unstable?

mkevins replied:
Good idea. Thanks for the suggestion!

geriux replied:
Updated https://github.com/WordPress/gutenberg/pull/29772/commits/756c056215227fa1b2b3ae99caed4e6e25807142

Commit:
Mark AutocompletionItemsSlot and AutocompletionItemsFill as __unstable

Commit:
Update styles to BEM naming

Commit:
Merge branch 'trunk' into rnmobile/feature/slash-inserter

# Conflicts:
#	packages/react-native-editor/CHANGELOG.md

Commit:
Revert "Use I with serif for italic format library icon"

This reverts commit c251037a457709d0f4919863ccd9be3c3d7a3929.

Commit:
Fix slash inserter e2e

Commit:
Fix e2e test on Android

enejb commented on packages/components/src/autocomplete/autocompleter-ui.native.js:
```diff
@@ -0,0 +1,233 @@
+
+const { compose: stylesCompose } = StyleSheet;
+
+const BackgroundView = ( { children } ) => {
```
NIT: This could be its own component file that has 2 different files one for iOS and another for Android. 
We could do this refactor in a different PR.

enejb commented on packages/edit-post/src/components/layout/index.native.js:
```diff
@@ -150,11 +151,17 @@ class Layout extends Component {
 						<BottomSheetSettings />
 					</KeyboardAvoidingView>
 				) }
+				{ Platform.OS === 'android' && <AutocompletionItemsSlot /> }
```
Why do we have this difference between iOS and Android?

mkevins replied:
If I recall correctly, this is due to differences in the placement of the `FloatingToolbar` component on iOS vs. Android.

enejb commented on packages/react-native-editor/__device-tests__/gutenberg-editor-slash-inserter-@canary.test.js:
```diff
@@ -0,0 +1,145 @@
+		await editorPage.driver.sleep( ANIMATION_TIME );
+
+		// Slash inserter
+		const slashInserterElement = await editorPage.driver.elementByAccessibilityId(
```
This code repeats itself a few times, do you think it would be helpful if we create a function for it? 

We could have a `assertSlashInserterPresent` and `assertSlashInserterNotPresent` methods for example.

enejb commented:
Nice work team! 

I tested this and it works really well. One thing I noticed is that if you mistype a block the autocomplete bar disappears right away (No animation). This is a really minor thing. 

I left a couple of minor notes and a few questions. (Feel free to ignore them :) ) 

I am happy to test this again after the rebase and give it the final :+1:

enejb commented on packages/components/src/autocomplete/style.android.scss:
```diff
@@ -0,0 +1,7 @@
+@import "./style.native.scss";
+
+.components-autocomplete {
+	width: 100%;
```
> width: 100%;
Is this needed?

mkevins committed:
Extract slash inserter assertion into separate function

mkevins committed:
Extract platform variants for BackgroundView to separate files

Commit:
Merge branch 'trunk' into rnmobile/feature/slash-inserter

# Conflicts:
#	packages/components/src/autocomplete/index.js
#	packages/react-native-editor/CHANGELOG.md

enejb reviewed (approved):
This worked well for me. And the code looks great! 
Nice work team!

Commit:
Merge branch 'trunk' into rnmobile/feature/slash-inserter

Commit:
Merge branch 'trunk' into rnmobile/feature/slash-inserter

Commit:
Merge branch 'trunk' into rnmobile/feature/slash-inserter

Commit:
Merge branch 'trunk' into rnmobile/feature/slash-inserter
